aiidalab_widgets_base/structures.py

```diff
@@ -99,59 +149,27 @@ def refresh_view(self):
         # pylint: disable=protected-access
         for comp_id in viewer._ngl_component_ids:
             viewer.remove_component(comp_id)
-
-        viewer.add_component(nglview.ASEStructure(
-            self.structure_ase))  # adds ball+stick
+        if self.structure_ase is None:
+            return
+        viewer.add_component(nglview.ASEStructure(self.structure_ase))  # adds ball+stick
         viewer.add_unitcell()
 
     # pylint: disable=unused-argument
     def _on_click_store(self, change):
-        self.store_structure(
-            self.file_upload.filename,
-            description=self.structure_description.value)
+        self.store_structure()
 
-    def store_structure(self, name, description=None):
-        structure_ase = self.get_ase(self.tmp_folder + '/' + name)
-        if structure_ase is None:
+    def store_structure(self, label=None, description=None):
+        if self.structure_node is None:
             return
-
-        # determine data source
-        if name.endswith('.cif'):
-            source_format = 'CIF'
-        else:
-            source_format = 'ASE'
-
-        # perform conversion
-        if self.data_format.value == 'CifData':
-            if source_format == 'CIF':
-                from aiida.orm.data.cif import CifData
-                structure_node = CifData(
-                    file=self.tmp_folder + '/' + name,
-                    scan_type='flex',
-                    parse_policy='lazy')
-            else:
-                from aiida.orm.data.cif import CifData
-                structure_node = CifData()
-                structure_node.set_ase(structure_ase)
-        else:  # Target format is StructureData
-            from aiida.orm.data.structure import StructureData
-            structure_node = StructureData(ase=structure_ase)
-
-            #TODO: Figure out whether this is still necessary for StructureData
-            # ensure that tags got correctly translated into kinds
-            for t1, k in zip(structure_ase.get_tags(),
-                             structure_node.get_site_kindnames()):
-                t2 = int(k[-1]) if k[-1].isnumeric() else 0
-                assert t1 == t2
-        if description is None:
-            structure_node.description = self.get_description(
-                structure_ase, name)
-        else:
-            structure_node.description = description
-        structure_node.label = ".".join(name.split('.')[:-1])
-        structure_node.store()
-        self.structure_node = structure_node
-        print("Stored in AiiDA: " + repr(structure_node))
+        if self.structure_node.is_stored:
+            print("Already stored in AiiDA: " + repr(self.structure_node) + " skipping..")
+            return
+        if label:
```

Comment:
so, since you are no longer passing the label in `_on_click_store`, will the node not have the correct label?

E.g. say you put label+desc. Then another widget asks for the node (=> takes label+desc).
Then you change the label in the widget & click "store".
Will the node still have the old  label?

Reply:
that is a bit the issue with this "on demand" design - you need to reset the cache whenever any field is edited that could affect the node

Reply:
solution: reset cache whenever any field that has effect on the node has changed.

Reply:
In the end I added `reset_structure` function that does only one thing: sets `self.structure_node` to `None`